### PR TITLE
Managed testing overhaul

### PR DESCRIPTION
**Pull request by mrvoorhe**
* The managed embedding tests in `UnityEmbedHost.Tests` will now run on our coreclr when ran from our CI script.

* Add `test.cmd` and `test.sh` scripts to make running just the `UnityEmbedHost.Tests` from the command line easier.

* When running the managed embedding tests from the CI scripts, our GC will be used.  (I'm pretty sure)

    When I set `DOTNET_GCName` and ran `dotnet test` on the solution, CoreCLR crashed before ever running tests.  However, when I deferred setting `DOTNET_GCName` until the test process itself, there was no crash.

    I used process explorer to verify the env var was set in the test process.

    Deferring is done by using a run settings file to specify env vars for the test process.

* There are now C# unit tests that will test the native embedding api.

* PInvokes into the native embedding apis are source generated.

* Generate wrappers around the pinvokes to make them easier to use.

* `CoreCLRHostTestingWrappers` is gone.  This is now source generated.  It's essentially the same as the above so it was easy to get both with more or less the same source generation logic.

* Tests that call into our native embedding api are ignored when running from Rider.

## Loose Ends ##

macOs and Linux both crash initializing the runtime when our GC is used.  To avoid this, I decided not to set `DOTNET_GCName` and roll the dice.  I don't know if I'll circle back to trying to figure out what is wrong.  I'm hoping this is good enough to hold together until we move to make the embedding apis GC safe.  At which point we won't need our GC to make the tests GC safe.

I haven't figured out how to get Rider to use our coreclr to run tests.  Because of that, the test strategy is a bit unresolved.  Currently, I've dulpicated all of the tests.

If I can get Rider (and VS) to run tests on our coreclr, I think maybe we can drop the tests that directly call the managed callback methods.

If I can't get Ride (and VS) to nicely run tests on our coreclr, then the callback tests may need to stay so that there is at least something to run from the IDE.  If this happens, maybe I can somehow come up with some abstract that allows us to reuse 1 set of tests for both purposes.

## Other changes & Notes ##

* Formatting in csproj.  Rider kept wanting to change it.

* There was some refactoring to the source generator to clean things up as I added more to it.

* There was some refactoring to BuildDriver to clean things up as I added mroe to it.

### CODE REVIEW

File: unity/CITools/BuildDriver/GlobalConfig.cs [hunk 1]
```diff
@@ -8,5 +8,6 @@ public class GlobalConfig
     public required string Architecture;
     public required string Configuration;
     public required bool Silent;
+    public required TestTargets TestTargets;
```
Comment (UnityAlex) [inline unity/CITools/BuildDriver/GlobalConfig.cs]: I'm not a huge fan of adding TestTargets here just for the embedding layer tests. I'd rather we change `Test` to `TestManaged` and then call them back to back based on what flags are set from `Program`

File: unity/CITools/BuildDriver/TestTargets.cs [hunk 2]
```diff
@@ -7,7 +7,9 @@ namespace BuildDriver;
 public enum TestTargets
 {
     None = 1 << 0,
```
Comment (UnityAlex) [inline unity/CITools/BuildDriver/TestTargets.cs]:
```suggestion
    None = 0,
```

File: unity/CITools/BuildDriver/TestTargets.cs [hunk 3]
```diff
@@ -7,7 +7,9 @@ namespace BuildDriver;
...
-    Embedding = 1 << 1,
-    CoreClr = 1 << 2,
+    EmbeddingManaged = 1 << 2,
```
Comment (UnityAlex) [inline unity/CITools/BuildDriver/TestTargets.cs]:
```suggestion
    EmbeddingManaged = 1 << 1,
```

File: unity/CITools/BuildDriver/TestTargets.cs [hunk 4]
```diff
@@ -7,7 +7,9 @@ namespace BuildDriver;
...
+    EmbeddingNative = 1 << 3,
```
Comment (UnityAlex) [inline unity/CITools/BuildDriver/TestTargets.cs]:
```suggestion
    EmbeddingNative = 1 << 2,
```

File: unity/CITools/BuildDriver/TestTargets.cs [hunk 5]
```diff
@@ -7,7 +7,9 @@ namespace BuildDriver;
...
+    Embedding = EmbeddingNative | EmbeddingManaged,
+    CoreClr = 1 << 4,
```
Comment (UnityAlex) [inline unity/CITools/BuildDriver/TestTargets.cs]:
```suggestion
    CoreClr = 1 << 3,
```

**Review by UnityAlex — CHANGES_REQUESTED** [review]

File: unity/CITools/BuildDriver/GlobalConfig.cs (hunk 1 above)
Comment (mrvoorhe) [reply unity/CITools/BuildDriver/GlobalConfig.cs]: Ok.  I'll make that change.

File: unity/CITools/BuildDriver/TestTargets.cs (hunk 2 above)
Comment (mrvoorhe) [reply unity/CITools/BuildDriver/TestTargets.cs]: Done

File: unity/CITools/BuildDriver/TestTargets.cs (hunk 3 above)
Comment (mrvoorhe) [reply unity/CITools/BuildDriver/TestTargets.cs]: Good catch.  Fixed

File: unity/CITools/BuildDriver/TestTargets.cs (hunk 4 above)
Comment (mrvoorhe) [reply unity/CITools/BuildDriver/TestTargets.cs]: Done

File: unity/CITools/BuildDriver/TestTargets.cs (hunk 5 above)
Comment (mrvoorhe) [reply unity/CITools/BuildDriver/TestTargets.cs]: Done

File: unity/CITools/BuildDriver/TestTargets.cs [hunk 6]
```diff
@@ -6,8 +6,10 @@ namespace BuildDriver;
 [Flags]
 public enum TestTargets
 {
-    None = 1 << 0,
-    Embedding = 1 << 1,
-    CoreClr = 1 << 2,
+    None = 0,
+    EmbeddingManaged = 1 << 1,
```
Comment (UnityAlex) [inline unity/CITools/BuildDriver/TestTargets.cs]: nit: I think you want to start at `1 << 0` not `1 << 1` Sorry if I messed this up in my original comment. That way you start at 1 and not 2.

**Review by UnityAlex — APPROVED** [review]

File: src/coreclr/dlls/mscoree/mscorwks_unixexports.src [hunk 7]
```diff
@@ -15,6 +15,7 @@ g_dacTable
 MetaDataGetDispenser
 
 ; mono
+initialize_unity
```
Comment (joncham) [inline src/coreclr/dlls/mscoree/mscorwks_unixexports.src]: Not a blocker, but could be named better based on both convention and purpose. Maybe `mono_unity_initialize_host_apis`?

**Review by joncham — APPROVED** [review]

File: src/coreclr/dlls/mscoree/mscorwks_unixexports.src (hunk 7 above)
Comment (mrvoorhe) [reply src/coreclr/dlls/mscoree/mscorwks_unixexports.src]: Done.

**Pull request merged**